config/puma/production.rb

```diff
@@ -17,3 +22,6 @@
   dynflow = ::Rails.application.dynflow
   dynflow.initialize! unless dynflow.config.lazy_initialization
 end
+
+# === Puma control rack application ===
+activate_control_app "unix://#{run_dir}/sockets/pumactl.sock"
```

Comment:
I haven't checked the permissions on this, but I'm assuming it's `foreman:foreman` and `0644` which should prevent from everyone writing to it. Correct?

Reply:
This is how it looks now:

# ll -d /run/foreman/ /run/foreman/sockets
drwxr-xr-x. 5 foreman foreman 200 Apr 22 14:30 /run/foreman/
drwxr-xr-x. 2 foreman foreman  60 Apr 22 14:23 /run/foreman/sockets

# ll /run/foreman/sockets
total 0
srwxrwxrwx. 1 foreman foreman 0 Apr 22 14:23 pumactl.sock

Reply:
Does that mean that any user can connect to the socket and restart the server?

Reply:
Are these determined by packaging?

Reply:
yes, looks like a `postgres` user is able to restart the puma server if right `control_url` and `control_auth_token` is known. Any suggestions on how to lock this down further?
```console
# ll /var/run/foreman/puma.state 
-rw-r--r--. 1 foreman foreman 120 Apr 22 15:23 /var/run/foreman/puma.state

# cat /var/run/foreman/puma.state 
---
pid: 1672
control_url: unix:///run/foreman/sockets/pumactl.sock
control_auth_token: 6c445af0312fff70fe0f190a65644cf

# sudo -u postgres scl enable tfm 'pumactl --control-url 'unix:///var/run/foreman/sockets/pumactl.sock' --control-token 6c445af0312fff70fe0f190a65644cf  restart'
Command restart sent success
```

Reply:
https://github.com/puma/puma/blob/2668597ec1dd9546d83db9f2ec5ad092add483e6/lib/puma/state_file.rb#L11-L13 suggests it can't be configured to set permissions. Right now I don't have advise.

Reply:
What are the permissions on viewing the control token?

Reply:
see the first line of the post above. the control token is present in puma.state file which has permissions 644

Reply:
All I can think is having something to adjust the permissions after creation but really the answer is to get a change into Puma to allow configuring the state file permissions.

Reply:
> All I can think is having something to adjust the permissions after creation but really the answer is to get a change into Puma to allow configuring the state file permissions.

I am working to get a fix to the puma project.

Reply:
Submitted patch to puma project https://github.com/puma/puma/pull/2238 to address the state file security issue.